lib/ain-evm/src/evm.rs

```diff
@@ -442,77 +272,236 @@ impl EVMServices {
                 .send(Notification::Block(block.header.hash()))
                 .map_err(|e| format_err!(e.to_string()))?;
         }
-        self.core.tx_queues.remove(queue_id);
+        self.core.block_templates.remove(template_id);
         self.core.clear_account_nonce();
         self.core.clear_transaction_cache();
 
         Ok(())
     }
 
-    unsafe fn update_queue_state_from_tx(
+    unsafe fn update_block_template_state_from_tx(
         &self,
-        queue_id: u64,
-        tx: QueueTx,
-    ) -> Result<(SignedTx, U256, H256)> {
-        let (target_block, state_root, timestamp, total_gas_used, is_first_tx) = {
-            let queue = self.core.tx_queues.get(queue_id)?;
-
-            let state_root = queue.get_latest_state_root();
-            let data = queue.data.lock();
-
-            (
-                data.target_block,
-                state_root,
-                data.timestamp,
-                data.total_gas_used,
-                data.transactions.is_empty(),
-            )
-        };
+        template_id: u64,
+        tx: ExecuteTx,
+    ) -> Result<TxState> {
+        let block_template = self.core.block_templates.get(template_id)?;
+        let state_root = block_template.get_latest_state_root();
+        let mut logs_bloom = block_template.get_latest_logs_bloom();
         debug!(
-            "[update_queue_state_from_tx] state_root : {:#?}",
+            "[update_block_template_state_from_tx] state_root : {:#?}",
             state_root
         );
 
+        let template = block_template.data.lock();
         let mut backend = EVMBackend::from_root(
             state_root,
             Arc::clone(&self.core.trie_store),
             Arc::clone(&self.storage),
-            Vicinity {
-                timestamp: U256::from(timestamp),
-                block_number: target_block,
-                block_gas_limit: U256::from(
-                    self.storage.get_attributes_or_default()?.block_gas_limit,
-                ),
-                ..Vicinity::default()
-            },
+            template.vicinity.clone(),
         )?;
+        let mut executor = AinExecutor::new(&mut backend);
 
+        let (parent_hash, _) = self
+            .block
+            .get_latest_block_hash_and_number()?
+            .unwrap_or_default(); // Safe since calculate_base_fee will default to INITIAL_BASE_FEE
+        let base_fee = self.block.calculate_base_fee(parent_hash)?;
```

Comment:
We can store base_fee in block template as well

Reply:
Nice spot, saw it as well, fixed in commit ded07f5 which also adds block difficulty into the template.